### PR TITLE
[SYCL] Fix various runtime compilation warnings

### PR DESCRIPTION
**Pull request by sergey-semenov**
- implicit conversion of string to bool
- unused variables
- default label in switch which covers all enumeration items
- unannotated fall-through in switch

### CODE REVIEW

**Review by kbobrovs — APPROVED** [review]
program_manager part looks OK. The signature change should not break ABI, AAICS.

File: sycl/source/detail/pi.cpp [hunk 1]
```diff
@@ -142,10 +142,9 @@ std::string platformInfoToString(pi_platform_info info) {
     return "PI_PLATFORM_INFO_VENDOR";
   case PI_PLATFORM_INFO_EXTENSIONS:
     return "PI_PLATFORM_INFO_EXTENSIONS";
-  default:
-    die("Unknown pi_platform_info value passed to "
-        "cl::sycl::detail::pi::platformInfoToString");
   }
+  die("Unknown pi_platform_info value passed to "
```
Comment (smaslov-intel) [inline sycl/source/detail/pi.cpp:146]: NIT: This is strange that compiler didn't see this unreachable (if I correctly understand what warning you are trying to fix). But, don't we have a coding standard that each "switch" should have a "default"?

**Review by smaslov-intel — APPROVED** [review]

**Review by v-klochkov — APPROVED** [review]

File: sycl/include/CL/sycl/detail/defines_elementary.hpp [hunk 2]
```diff
@@ -48,3 +48,24 @@
...
+
+#ifndef __SYCL_HAS_CPP_ATTRIBUTE
+#if defined(__cplusplus) && defined(__has_cpp_attribute)
+#define __SYCL_HAS_CPP_ATTRIBUTE(x) __has_cpp_attribute(x)
+#else
+#define __SYCL_HAS_CPP_ATTRIBUTE(x) 0
+#endif
+#endif
+
+#if defined(__cplusplus) && __cplusplus > 201402L &&                           \
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:60]: late nit:
```suggestion
#ifndef __SYCL_FALLTHROUGH
#if defined(__cplusplus) && __cplusplus > 201402L &&                           \
```

File: sycl/include/CL/sycl/detail/defines_elementary.hpp [hunk 3]
```diff
@@ -48,3 +48,24 @@
...
+#define __SYCL_FALLTHROUGH [[fallthrough]]
+#elif __SYCL_HAS_CPP_ATTRIBUTE(gnu::fallthrough)
+#define __SYCL_FALLTHROUGH [[gnu::fallthrough]]
+#elif __has_attribute(fallthrough)
+#define __SYCL_FALLTHROUGH __attribute__((fallthrough))
+#elif __SYCL_HAS_CPP_ATTRIBUTE(clang::fallthrough)
+#define __SYCL_FALLTHROUGH [[clang::fallthrough]]
+#else
+#define __SYCL_FALLTHROUGH
+#endif
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:71]: late nit:
```suggestion
#endif
#endif // __SYCL_FALLTHROUGH
```

File: sycl/source/detail/memory_manager.cpp [hunk 4]
```diff
@@ -47,6 +47,8 @@ void MemoryManager::release(ContextImplPtr TargetContext, SYCLMemObjI *MemObj,
 
 void MemoryManager::releaseImageBuffer(ContextImplPtr TargetContext,
                                        void *ImageBuf) {
+  (void)TargetContext;
```
Comment (kbobrovs) [inline sycl/source/detail/memory_manager.cpp:50]: late nit: suggest to introduce macros like
`#define UNUSED_ARGUMENT(x) (void)x`
`#define UNUSED_ARGUMENTS2(x,y) (void)x; (void) y`
 and them instead.

File: sycl/source/detail/pi.cpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/source/detail/pi.cpp:146]: I believe this fix here is for -Wcovered-switch-default warning, i.e. when all cases in switch are listed, but the default statement is still there (i.e. redundant).

**Review by kbobrovs — APPROVED** [review]

File: sycl/source/detail/memory_manager.cpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/source/detail/memory_manager.cpp:50]: Please not in this patch as we already have hundreds of places using this old style "(void)TargetContext".
It should be a separate PR adding those macros, imo.

File: sycl/include/CL/sycl/detail/defines_elementary.hpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:78]: Fixed in https://github.com/intel/llvm/pull/2980/commits/a60e63f1c597cd83e588fe0cf5337929998ba83f

File: sycl/include/CL/sycl/detail/defines_elementary.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:67]: Fixed in https://github.com/intel/llvm/pull/2980/commits/a60e63f1c597cd83e588fe0cf5337929998ba83f

File: sycl/source/detail/pi.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/source/detail/pi.cpp:146]: OK, here it is safe to dies after the switch.

**Review by smaslov-intel — APPROVED** [review]
pi.cpp changes LGTM

File: sycl/source/detail/memory_manager.cpp (hunk 4 above)
Comment (kbobrovs) [reply sycl/source/detail/memory_manager.cpp:50]: Separate patch sounds good to me.

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**